File: HW1/calculate_ellipse.m
function [Xe,Ye,U,S,th] = calculate_ellipse(X, P, nsig, np) 
% This functions returns points to draw a 2D ellipse
%
%  [Xe,Ye,U,S,th] = calculate_ellipse(X, P, nsig, np)
% 
% INPUTS
%  @param X     2D mean 
%  @param P     2D covariance matrix 
%  @param nsig  number of sigmas (default is 1) 
%  @param np    number of plot points (default is 36) 
% 
% OUTPUTS
%  @param Xe    ellipse x coordinate
%  @param Ye    ellipse x coordinate
%  @param U     U from svd of covariance (as a rotation matrix)
%  @param S     S from svd of covariance
%  @param th    rotation angle
%
% MAE 6760 Model Based Estimation
% Cornell University
% M Campbell
%
error(nargchk(2, 4, nargin)); 
if nargin<3, nsig = 1; end 
if nargin<4, np = 36; end 

if (length(X) ~= 2) | (length(P) ~= 2), 
    disp('mean and covariance have to be 2D');
    Xe=[];Ye=[];U=[];S=[];th=[];
    return;
end

[U,S,V]=svd(P); %svd of the covariance
%make the U matrix a true rotation matrix: [cos(th) -sin(th);sin(th) cos(th)
if U(1,1) ~= 0, %just be sure diagonal is not all zeros
    U=U*diag(sign(diag(U)));
end
%find rotation angle
th=-asin(U(1,2))*180/pi;

s1=sqrt(S(1,1));s2=sqrt(S(2,2));
x=X(1);
y=X(2);

%scale by nsig
s1=nsig*s1;
s2=nsig*s2;

beta = th * (pi / 180); 
sinbeta = sin(beta); 
cosbeta = cos(beta); 

alpha = linspace(0, 360, np)' .* (pi / 180); 
sinalpha = sin(alpha); 
cosalpha = cos(alpha); 

Xe = x + (s1 * cosalpha * cosbeta - s2 * sinalpha * sinbeta); 
Ye = y + (s1 * cosalpha * sinbeta + s2 * sinalpha * cosbeta); 
 
end

File: HW4/calculate_ellipse.py
import numpy as np

def calculate_ellipse(X, P, nsig=1, np_points=36):
    """
    Calculate points of a 2D covariance ellipse.

    Parameters
    ----------
    X : array-like, shape (2,)
        2D mean vector
    P : array-like, shape (2,2)
        2D covariance matrix
    nsig : float, optional
        Number of sigmas (default = 1)
    np_points : int, optional
        Number of ellipse points (default = 36)

    Returns
    -------
    Xe : ndarray
        Ellipse x-coordinates
    Ye : ndarray
        Ellipse y-coordinates
    U : ndarray
        Left singular vectors (rotation matrix)
    S : ndarray
        Singular value matrix
    th : float
        Rotation angle in degrees

    MAE 6760 Model Based Estimation
    Cornell University
    M Campbell
    """

    X = np.asarray(X).reshape(-1)
    P = np.asarray(P)

    if X.size != 2 or P.shape != (2, 2):
        raise ValueError("Mean and covariance must be 2D")

    # SVD of covariance
    U, S_vals, Vt = np.linalg.svd(P)
    S = np.diag(S_vals)

    # Make U a true rotation matrix (match MATLAB sign correction)
    if U[0, 0] != 0:
        U = U @ np.diag(np.sign(np.diag(U)))

    # Rotation angle (degrees)
    th = -np.arcsin(U[0, 1]) * 180.0 / np.pi

    # Semi-axis lengths
    s1 = nsig * np.sqrt(S[0, 0])
    s2 = nsig * np.sqrt(S[1, 1])

    x, y = X

    # Angle definitions
    beta = np.deg2rad(th)
    sinbeta = np.sin(beta)
    cosbeta = np.cos(beta)

    alpha = np.linspace(0, 360, np_points) * np.pi / 180.0
    sinalpha = np.sin(alpha)
    cosalpha = np.cos(alpha)

    # Ellipse parametric equations
    Xe = x + (s1 * cosalpha * cosbeta - s2 * sinalpha * sinbeta)
    Ye = y + (s1 * cosalpha * sinbeta + s2 * sinalpha * cosbeta)

    return Xe, Ye, U, S, th